### PR TITLE
Issue 907

### PR DESCRIPTION
**Pull request by vadimeisenbergibm**
https://github.com/IBM-Swift/Kitura/issues/907
@ianpartridge please review and merge

### CODE REVIEW

File: Sources/KituraSampleRouter/RouterCreator.swift [hunk 1]
```diff
@@ -21,6 +21,7 @@ import Foundation
 import Kitura
 import KituraMustache
 import KituraStencil
+import Stencil // required for adding a namespace to Stencil Template Engine
```
Comment (ianpartridge) [inline Sources/KituraSampleRouter/RouterCreator.swift]: GitHub says there's a problem with this line?

File: Package.swift [hunk 2]
```diff
@@ -27,6 +27,7 @@ let package = Package(
         .Package(url: "https://github.com/IBM-Swift/Kitura.git", majorVersion: 1, minor: 3),
         .Package(url: "https://github.com/IBM-Swift/HeliumLogger.git", majorVersion: 1, minor: 3),
         .Package(url: "https://github.com/IBM-Swift/Kitura-MustacheTemplateEngine.git", majorVersion: 1, minor: 1),
-        .Package(url: "https://github.com/IBM-Swift/Kitura-StencilTemplateEngine.git", majorVersion: 1, minor: 3),
+        .Package(url: "https://github.com/IBM-Swift/Kitura-StencilTemplateEngine.git", majorVersion: 1, minor: 4),
+        .Package(url: "https://github.com/kylef/Stencil", majorVersion: 0, minor: 7)
```
Comment (ianpartridge) [inline Package.swift]: Is it necessary to import Stencil as well? Doesn't StencilTemplateEngine give us this?

File: Sources/KituraSampleRouter/RouterCreator.swift (hunk 1 above)
Comment (vadimeisenbergibm) [inline Sources/KituraSampleRouter/RouterCreator.swift]: My guess is that github does not like a whitespace before the comment //

Comment (ianpartridge) [inline Sources/KituraSampleRouter/RouterCreator.swift]: More likely the whitespace is a non-ASCII character?

File: Sources/KituraSampleRouter/RouterCreator.swift [hunk 3]
```diff
@@ -20,7 +20,8 @@ import Foundation
 
 import Kitura
 import KituraMustache
-import KituraStencil
+import KituraStencil // required for using StencilTemplateEngine
```
Comment (vadimeisenbergibm) [inline Sources/KituraSampleRouter/RouterCreator.swift]: @ianpartridge this is the whitespace on my keyboard. Could you please the following check for me - checkout the branch https://github.com/IBM-Swift/Kitura-Sample/tree/issue_907, remove these whitespaces and add them, perform `git diff` and see if the whitespaces on your keyboard are different?

Comment (ianpartridge) [inline Sources/KituraSampleRouter/RouterCreator.swift]: I tried this and `git diff` showed no difference. Perhaps it is  GitHub anomaly.

File: Package.swift (hunk 2 above)
Comment (vadimeisenbergibm) [inline Package.swift]: In C/C++, it is considered a good practice by some not to relied on headers included in other headers, and to include all the required headers by some file in that file. See http://softwareengineering.stackexchange.com/a/262020. I am not sure regarding what is the best practice in SPM, my reasoning is that it makes dependency explicit - Kitura-Sample code is dependent on Stencil package directly, specifically on `Namespace` class of `Stencil` package.

Comment (youming-lin) [inline Package.swift]: Explicit dependency that duplicate a sub-dependency can cause dependency graph errors in SPM if the versions do not match (unlike `npm` which can manage different versions of the same package). As such I prefer to keep `Package.swift` as minimal as possible.

Perhaps the solution here would be to leave the Stencil package as a comment as opposed to an actual `.Package`?

Comment (vadimeisenbergibm) [inline Package.swift]: @youming-lin I still do not see what is the problem. Suppose you have module A that depends on module B and module C, and module B also depends on module C. You state in Package.swift of module A which version of C you expect, and the author of the module B specifies in its Package.swift which version of C module B expects. SPM will fail if the expectations of the versions are different - you get additional check. What is wrong with that?

Comment (youming-lin) [inline Package.swift]: The additional check doesn't provide any additional safeguard (module A's unit tests should already guarantee functional correctness), but would necessitate an update to module A, even if the module C update doesn't create any functional regression, and even if no code change was made to module A, simply because an update to module B causes a new module C to be pulled in and cause dependency graph error in module A.

TL;DR: Explicitly specifying module C in module A's `Package.swift` can cause breakage in module A even if no change was made to module A and no breaking change was introduced in module C.

Comment (vadimeisenbergibm) [inline Package.swift]: > breakage in module A even if no change was made to module A and no breaking change was introduced in module C.
In this case, the version of C in module A's `Package.swift` should be updated and no harm will be caused.

Consider an alternative case where a change in B makes B dependent on a version of C which is incompatible with A. In this case, if you do not have a dependency on C in module A's `Package.swift` you will get a linker error (for example, that some function in C that A expects, is missing.) Instead of getting a clear message that you have versions conflict, you will perform full compilation and get linker error instead, which will be harder to understand. Especially if you compile D that depends on A that depends on B and C. You will have to chase where this linker error came from instead of having clear version conflict error from SPM.

**Comment by shmuelk** [comment]
Version conflicts in SPM are NOT clear at all. You don't know who is requesting which version of what.

@ylin is correct.

**Comment by vadimeisenbergibm** [comment]
@shmuelk so do you prefer to get a linker error instead of SPM error regarding the version conflict? With SPM version conflict error you at least have a chance to find out where is the conflict, e.g. by performing grep on `Package.swift` files in the sub-directories of `Packages` directory. It should be easier than to understand a linker error.

File: Package.swift (hunk 2 above)
Comment (youming-lin) [inline Package.swift]: In an ideal world, as long as module A's `Package.swift` follows SemVer, and modules B and C, as long as they follow SemVer when tagging releases, will not introduce a breaking change that would cause module A to break.

However, if you explicitly include module C in module A's `Package.swift`, even if a non-breaking update to module C was released, module A might break WITHOUT any change in module A's source code whatsoever.

This, to me, is a big no-no. We don't want, say, a tagged release of module A to suddenly break simply because of a dependency graph error introduced by an update to a dependency that both module A and module C rely on.

How could this happen?
Say module A is tagged 1.2.0, and relies on version 1.5.x of module B. Version 1.5.1 of module B depends on version 2.3.x of module C. Module C just got updated to version 2.4.0 from 2.3.7 (this IS NOT a breaking API change according to SemVer). Then, module B is updated to version 1.5.2 to pull in version 2.4.0 of module C.

Now, version 1.2.0 of module A is broken because of dependency graph error due to module C 2.3.7->2.4.0 update. This means that all projects that rely on version 1.2.0 of module A is now suddenly broken even though there has been no change to module A whatsoever, and that the module C update is a non-breaking API change according to SemVer.

TL;DR: Having redundant packages listed in `Package.swift` can break an existing, tagged release through no fault of its own.

Comment (vadimeisenbergibm) [inline Package.swift]: @youming-lin I think I understand the issue, the change in your example from 2.3 to 2.4 should not break the APIs and module A should not be affected. Omitting a dependency in module A's `Package.swift` practically says: module A does not care which version of module C is fetched due to the dependency in module B's `Package.swift`, if this dependency does not introduce a breaking change in APIs of module C, i.e. the version of module C does not change to 3.0.0. But this can be expressed by specifying the version of module C as 2.x.x (only the major version). 

So how about "saying what you mean" and specifying the major version only? Nothing will be broken until the major version will be increased.

This would work:
```swift
.Package(url: "https://github.com/kylef/Stencil", majorVersion: 0)
```

Comment (youming-lin) [inline Package.swift]: IIRC SemVer allows for breaking changes on minor version change if the major version number is 0, like in your example.

http://semver.org/#spec-item-4

If a new update to module C breaks module A due to an API change, the compiler should be able to point out the offending API call, and from there the programmer should be able to see that the external API to module C has changed, causing this compile error. After that it's easy to guess that a version update to module C was the cause of this compile error.

Comment (vadimeisenbergibm) [inline Package.swift]: > IIRC SemVer allows for API-breaking changes on minor version change if the major version number is 0, like in your example.

Hmm, so in case of major version 0, increasing a minor version in the version of C can introduce a breaking API change. So you think that instead of having a version conflict immediately, it is better to proceed with fetching and compilation, until the compiler error?

Still, so what harm will arise from specifying only the major version?
```swift
.Package(url: "https://github.com/kylef/Stencil", majorVersion: 0)
```
Only when Kitura-TemplateEngine will switch to Stencil 1.0.0, a version conflict will happen, and it will force the author of Kitura-Sample to check the differences in the APIs of Stencil and to update the major version to 1.

Comment (youming-lin) [inline Package.swift]: I feel that, module A should only break IF AND ONLY IF the update to module C is API breaking. Otherwise, as long as a version update for module C doesn't break the relevant API that is used, even if this version update is a major version update, module A shouldn't break.

Again, for API changes, compiler should catch them. For underlying functionality changes (same API but different behavior), unit tests in module A should catch them.

However, in all honesty, a tagged release of module A shouldn't ever break, implicit dependency version updates notwithstanding, if it had already passed tests before being tagged for release, since module A is supposed to contain the full stack of dependencies required to compile it.


Another side to this thinking is: why don't we nail down the specific version of module B, so that for a given tagged release of module A, we would NEVER pull in a different version of module B, not even a minor update?

I think this discussion highlights some downsides to the way SPM handles versioning that really shouldn't happen in a production environment, and hardening versions down to the minor-est of version numbers across the entire dependency stack seems to be the only way to prevent these pitfalls from happening.

Comment (youming-lin) [inline Package.swift]: Also, I like to add that SPM versioning issues (not this one specifically) has caused a few tagged releases breaking in the past. [This issue](https://github.com/IBM-Swift/Kitura/issues/874) comes to mind.

Comment (vadimeisenbergibm) [inline Package.swift]: > Again, for API changes, compiler should catch them. For underlying functionality changes (same API but different behavior), unit tests in module A should catch them.

If so, I do not understand why do you need versioning at all. You can just specify no version for all the dependencies, and compiler/unit tests would catch any problems due to version mismatches.

**Comment by shmuelk** [comment]
Please remove the extra dependency and we will merge this PR.
